### PR TITLE
Feature test runner procedure

### PR DESCRIPTION
**Pull request by mbentley124**

### CODE REVIEW

**Comment by ndalporto** [comment]
>tfw david reviews loeffler's pull request and not your

File: src/org/team5940/robot_core/modules/testing/runners/MultipleTestRunnerProcedureModule.java [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+import org.team5940.robot_core.modules.control.procedures.AbstractProcedureModule;
+import org.team5940.robot_core.modules.logging.LoggerModule;
+import org.team5940.robot_core.modules.testing.TestableModule;
+import org.team5940.robot_core.modules.testing.TestableModule.TestStatus;
+import org.team5940.robot_core.modules.testing.communication.TestCommunicationModule;
+
+import com.sun.org.apache.bcel.internal.generic.Type;
+
+/**
+ * A {@link AbstractProcedureModule} that runs tests on {@link TestableModule}.
```
Comment (david-boles) [inline src/org/team5940/robot_core/modules/testing/runners/MultipleTestRunnerProcedureModule.java]: .. runs tests on on +a ...

File: src/org/team5940/robot_core/modules/testing/runners/MultipleTestRunnerProcedureModule.java [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+				new Object[] { testableModules, communicationModule });
+		this.testableModules = testableModules;
+		this.communicationModule = communicationModule;
+		this.logger.logInitialization(this, MultipleTestRunnerProcedureModule.class,
+				new Object[] { testableModules, communicationModule });
+	}
+
+	@Override
+	protected synchronized void doProcedureStart() throws Exception {
+		testThread = new TestThread(this.communicationModule, this.testableModules);
```
Comment (david-boles) [inline src/org/team5940/robot_core/modules/testing/runners/MultipleTestRunnerProcedureModule.java:66]: Interrupt the old thread if not null.

File: src/org/team5940/robot_core/modules/testing/runners/MultipleTestRunnerProcedureModule.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+	}
+
+	@Override
+	protected synchronized boolean doProcedureUpdate() throws Exception {
+		return testThread.isInterrupted();
+	}
+
+	@Override
+	protected synchronized void doProcedureClean() {
+		testThread.interrupt();
```
Comment (david-boles) [inline src/org/team5940/robot_core/modules/testing/runners/MultipleTestRunnerProcedureModule.java:77]: Reset to null

File: src/org/team5940/robot_core/modules/testing/runners/MultipleTestRunnerProcedureModule.java (hunk 2 above)
Comment (david-boles) [reply src/org/team5940/robot_core/modules/testing/runners/MultipleTestRunnerProcedureModule.java:66]: NVM

File: src/org/team5940/robot_core/modules/testing/runners/MultipleTestRunnerProcedureModule.java [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+ * extend/implement.
+ * 
+ * @author Michael Bentley
+ */
+public class MultipleTestRunnerProcedureModule extends AbstractProcedureModule {
+
+	/**
+	 * The modules to test.
+	 */
+	ModuleHashtable<TestableModule> testableModules;
```
Comment (david-boles) [inline src/org/team5940/robot_core/modules/testing/runners/MultipleTestRunnerProcedureModule.java]: Make all variables private and except for thread final

File: src/org/team5940/robot_core/modules/testing/runners/MultipleTestRunnerProcedureModule.java [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+	}
+
+	@Override
+	protected synchronized void doProcedureClean() {
+		testThread.interrupt();
+	}
+
+	private class TestThread extends Thread {
+
+		TestCommunicationModule communicationModule;
```
Comment (david-boles) [inline src/org/team5940/robot_core/modules/testing/runners/MultipleTestRunnerProcedureModule.java]: Same private and final variable dealio, document

File: src/org/team5940/robot_core/modules/testing/runners/MultipleTestRunnerProcedureModule.java [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+					}
+				}
+				this.communicationModule.displayText("Tests Complete");
+			} catch (InterruptedException e) {
+				e.printStackTrace();
+			}
+			this.interrupt();
+		}
+
+		private boolean isAssignableToString(String className, Class<?> module) {
```
Comment (david-boles) [inline src/org/team5940/robot_core/modules/testing/runners/MultipleTestRunnerProcedureModule.java]: Document

File: src/org/team5940/robot_core/modules/testing/runners/MultipleTestRunnerProcedureModule.java [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+			this.completed = false;
+		}
+
+		@Override
+		public void run() {
+			String testModuleName;
+			try {
+				TestableModule testModule;
+				String userActionString;
+				while (!this.completed) {
```
Comment (david-boles) [inline src/org/team5940/robot_core/modules/testing/runners/MultipleTestRunnerProcedureModule.java]: Switch to just using interruption

File: src/org/team5940/robot_core/modules/testing/runners/MultipleTestRunnerProcedureModule.java [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+							"Would you like to search, list, or test, the TestableModules? (Enter no to leave)");
+					if (userActionString.toLowerCase().equals("test")) {
+						do {
+							testModuleName = this.communicationModule
+									.promptText("Which module would you like to test?");
+							testModule = this.testableModules.get(testModuleName);
+							if (testModule == null) {
+								testModule = this.testableModules.get(testModuleName.toLowerCase());
+							}
+						} while (testModule == null);
```
Comment (david-boles) [inline src/org/team5940/robot_core/modules/testing/runners/MultipleTestRunnerProcedureModule.java:126]: Check for thread interruption.

File: src/org/team5940/robot_core/modules/testing/runners/MultipleTestRunnerProcedureModule.java [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+
+import org.team5940.robot_core.modules.Module;
+import org.team5940.robot_core.modules.ModuleHashtable;
+import org.team5940.robot_core.modules.control.procedures.AbstractProcedureModule;
+import org.team5940.robot_core.modules.logging.LoggerModule;
+import org.team5940.robot_core.modules.testing.TestableModule;
+import org.team5940.robot_core.modules.testing.TestableModule.TestStatus;
+import org.team5940.robot_core.modules.testing.communication.TestCommunicationModule;
+
+import com.sun.org.apache.bcel.internal.generic.Type;
```
Comment (david-boles) [inline src/org/team5940/robot_core/modules/testing/runners/MultipleTestRunnerProcedureModule.java]: Remove import

**Review by david-boles — CHANGES_REQUESTED** [review]

**Review by david-boles — APPROVED** [review]

**Pull request merged**